### DISCUSSION TITLE
Getting error when implementing partial_fit

### DISCUSSION THREAD

**Discussion by VenkataBuddha** [MaartenGr/BERTopic#944] [Q&A]
I am getting the following error: unsupported operand type(s) for -: 'NoneType' and 'NoneType'

I am using the following parametrs for Online Topic modelling:
# embeddings
embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
# Tokenize topics
vectorizer_model = OnlineCountVectorizer(stop_words="english", ngram_range=(1, 3))
# Create topic representation
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True, bm25_weighting=True)
cluster_model = River(cluster.TextClust())

# Prepare model
topic_model = BERTopic(embedding_model=embedding_model,                                  
                umap_model=umap_model,
        ctfidf_model=ctfidf_model,
    hdbscan_model=cluster_model, 
    vectorizer_model=vectorizer_model
)

**Comment by drob-xx** [comment]
Can you share the entire exception output?

**Reply by VenkataBuddha** [reply]
Sure:
TypeError                                 Traceback (most recent call last)
<command-3660042844608377> in <module>
      1 # Run BERTopic model
----> 2 topics1, probabilities1 = topic_model.partial_fit(df['Body_title_abs_lemmatized'])

/databricks/python/lib/python3.8/site-packages/bertopic/_bertopic.py in partial_fit(self, documents, embeddings, y)
    546 
    547         # Cluster reduced embeddings
--> 548         documents, self.probabilities_ = self._cluster_embeddings(umap_embeddings, documents, partial_fit=True)
    549         topics = documents.Topic.to_list()
    550 

/databricks/python/lib/python3.8/site-packages/bertopic/_bertopic.py in _cluster_embeddings(self, umap_embeddings, documents, partial_fit, y)
   2854         """
   2855         if partial_fit:
-> 2856             self.hdbscan_model = self.hdbscan_model.partial_fit(umap_embeddings)
   2857             labels = self.hdbscan_model.labels_
   2858             documents['Topic'] = labels

<command-3660042844608370> in partial_fit(self, umap_embeddings)
      8     def partial_fit(self, umap_embeddings):
      9         for umap_embedding, _ in stream.iter_array(umap_embeddings):
---> 10             self.model = self.model.learn_one(umap_embedding)
     11 
     12         labels = []

/databricks/python/lib/python3.8/site-packages/river/cluster/textclust.py in learn_one(self, x, t, sample_weight)
    196                 self.micro_clusters[clusterId].n += 1
    197 
--> 198                 self.micro_clusters[clusterId].merge(
    199                     mc, self.t, self.omega, self.fading_factor, self.term_fading, self.realtime
    200                 )

/databricks/python/lib/python3.8/site-packages/river/cluster/textclust.py in merge(self, microcluster, t, omega, fading_factor, term_fading, realtime)
    594             self.weight = self.weight + microcluster.weight
    595 
--> 596             self.fade(t, omega, fading_factor, term_fading, realtime)
    597             microcluster.fade(t, omega, fading_factor, term_fading, realtime)
    598 

/databricks/python/lib/python3.8/site-packages/river/cluster/textclust.py in fade(self, tnow, omega, fading_factor, term_fading, realtime)
    576         ## fading micro cluster weights and also term weights, if activated
    577         def fade(self, tnow, omega, fading_factor, term_fading, realtime):
--> 578             self.weight = self.weight * pow(2, -fading_factor * (tnow - self.time))
    579             if term_fading:

**Reply by drob-xx** [reply]
Can you show all the code from the point at which you `fit_transform` your BERTopic model to the point where you pop the error? What is `topic_model.umap_model.embedding_.shape`?

**Reply by VenkataBuddha** [reply]
I am actually starting of with my inital training : so i just ran the below line of code:

embedding_model = SentenceTransformer("all-MiniLM-L6-v2")

vectorizer_model = OnlineCountVectorizer(stop_words="english", ngram_range=(1, 3))

ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True, bm25_weighting=True)
cluster_model = River(cluster.TextClust())



topic_model = BERTopic(embedding_model=embedding_model,                                  
                umap_model=umap_model,
        ctfidf_model=ctfidf_model,
    hdbscan_model=cluster_model, 
    vectorizer_model=vectorizer_model
)

# Run BERTopic model
topics1, probabilities1 = topic_model.partial_fit(df['Body_title_abs_lemmatized'])

**Reply by drob-xx** [reply]
I'm not sure what is going on.

**Reply by MaartenGr** [reply]
From your code, I can see a number of things happening but the most important thing here is the cluster model that you are using. According to its [documentation](https://riverml.xyz/0.14.0/api/cluster/TextClust/), it expects a Bag-of-Words representation in the format that River is supplying which is not supported within BERTopic since it expects n-dimensional embeddings for each document and not a `collections.Counter` as is the case with River. According to its documentation, River expects the following type of pipeline:

```python
>>> model = compose.Pipeline(
...     feature_extraction.BagOfWords(lowercase=True, ngram_range=(1, 2), stop_words=stopwords),
...     cluster.TextClust(real_time_fading=False, fading_factor=0.001, tgap=100, auto_r=True,
...     radius=0.9)
... )`
```

In `BagOfWords` a `collections.Counter` is created which is fed into the `TextClust` algorithm, so you would have to update the internal coding of BERTopic in order to allow for these counters to be passed. Also, we will have to skip the dimensionality reduction step as that was not taken into account for this cluster model. The following is a working example of using `TextClust` in BERTopic:

```python
from tqdm import tqdm
from river import stream, cluster, feature_extraction, compose
from sklearn.datasets import fetch_20newsgroups

from bertopic import BERTopic
from bertopic.backend import BaseEmbedder
from bertopic.vectorizers import OnlineCountVectorizer, ClassTfidfTransformer
from bertopic.dimensionality import BaseDimensionalityReduction


# We create an embedding model out of River's BoW approach
class RiverBoW(BaseEmbedder):
    def __init__(self):
        super().__init__()

    def embed(self, documents, verbose=False):
        return documents 


# This function was updated to take in the counters one at a time
class River:
    def __init__(self):
        self.model = compose.Pipeline(
                        feature_extraction.BagOfWords(lowercase=True, ngram_range=(1, 2)),
                        cluster.TextClust(real_time_fading=False, fading_factor=0.001, tgap=100, auto_r=True, radius=0.9)
                        )

    def partial_fit(self, corpus):
        labels = []
        for x in tqdm(corpus):
            y_pred = self.model.predict_one(x)
            try:
              self.model = self.model.learn_one(x)
              if y_pred is None:
                labels.append(-1)
              else:
                labels.append(y_pred)
            except:
              labels.append(-1)
            
        self.labels_ = labels
        print(self.labels_)

        return self


# We skip creating topic vectors as they do not work well with non-numpy arrays
class BERTopicTextClust(BERTopic):
    def _create_topic_vectors(self):
      pass


# Prepare data and calculate bag-of-words beforehand
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

# Sub models
embedding_model = RiverBoW()
empty_dimensionality_model = BaseDimensionalityReduction()
cluster_model = River()
vectorizer_model = OnlineCountVectorizer(stop_words="english", ngram_range=(1, 3))
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True, bm25_weighting=True)

# Prepare topic model
topic_model = BERTopicTextClust(
    embedding_model=embedding_model,
    umap_model=empty_dimensionality_model,
    ctfidf_model=ctfidf_model,
    hdbscan_model=cluster_model,
    vectorizer_model=vectorizer_model
)

# Run topic model
topic_model.partial_fit(docs[:500])
```

Do note that there is quite a lot going on here and that diving into the River documentation might help understand some of the internal tricks that are happening here. Also, in order to support this some hacks need to be done which I typically do not recommend. Instead, following the example [here](https://maartengr.github.io/BERTopic/getting_started/online/online.html) might be preferred.
